src/test/java/com/artipie/asto/s3/InternalExceptionHandleTest.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 artipie.com
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto.s3;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Test for {@link InternalExceptionHandle}.
+ *
+ * @since 1.0
+ */
+final class InternalExceptionHandleTest {
+
+    @Test
+    void translatesException() {
+        final CompletableFuture<Void> future = CompletableFuture.runAsync(Assertions::fail);
+        MatcherAssert.assertThat(
+            Assertions.assertThrows(
+                ExecutionException.class,
+                future.handle(
+                    new InternalExceptionHandle<>(
+                        future,
+                        AssertionError.class,
+                        IllegalStateException::new
+                    )
+                )
+                    .thenCompose(Function.identity())
+                    .toCompletableFuture()
+                    ::get
+            ),
+            Matchers.hasProperty("cause", Matchers.isA(IllegalStateException.class))
+        );
+    }
```

Comment:
@andreoss nested assertions don't look good.  You can capture the exception in ` AtomicReference` and after that make required checks

Reply:
@genryxy `assertThrows` returns caught Exception especially for the later assertion

Reply:
@andreoss okay, got it